frontend/javascripts/oxalis/view/right-border-tabs/connectome_tab/connectome_view.js

```diff
@@ -0,0 +1,784 @@
+
+    // Auto-expand all nodes by default. The antd properties like `defaultExpandAll` only work on the first render
+    // but not when switching to another agglomerate, afterwards.
+    const treeData = convertConnectomeToTreeData(connectomeData) || [];
+    const expandedKeys = Array.from(
+      mapAndFilterTreeData(treeData, node => node.key, node => node.data.type !== "synapse"),
+    );
+
+    // Auto-load the skeletons of the active agglomerates and check all occurences of the same agglomerate
+    const topLevelCheckedKeys = treeData.map(topLevelTreeNode => topLevelTreeNode.key);
+    const checkedKeys = Array.from(
+      mapAndFilterTreeData(
+        treeData,
+        node => node.key,
+        node => topLevelCheckedKeys.some(topLevelKey => node.key.startsWith(topLevelKey)),
+      ),
+    );
+
+    this.setState({
+      connectomeData,
+      availableSynapseTypes: typeToString,
+      checkedKeys,
+      expandedKeys,
+    });
+  }
+
+  updateSynapseTrees(prevFilteredConnectomeData: ?ConnectomeData) {
+    const { segmentationLayer } = this.props;
+    const { filteredConnectomeData } = this.state;
+
+    if (segmentationLayer == null) return;
+
+    let prevFilteredSynapseIds: Array<number> = [];
+    let filteredSynapseIds: Array<number> = [];
+    if (prevFilteredConnectomeData != null) {
+      prevFilteredSynapseIds = getSynapseIdsFromConnectomeData(prevFilteredConnectomeData);
+    }
+    if (filteredConnectomeData != null) {
+      filteredSynapseIds = getSynapseIdsFromConnectomeData(filteredConnectomeData);
+    }
+
+    const reset =
+      prevFilteredConnectomeData != null &&
+      filteredConnectomeData != null &&
+      prevFilteredConnectomeData.connectomeFile !== filteredConnectomeData.connectomeFile;
+
+    let deletedSynapseIds;
+    let addedSynapseIds;
+    if (reset) {
+      // If the data needs to be reset, because the connectome file has changed, all existing trees need to be removed
+      // and all non-existing trees need to be newly added. Otherwise, IDs of one connectome file would get mixed up
+      // with IDs from the other.
+      deletedSynapseIds = prevFilteredSynapseIds;
+      addedSynapseIds = filteredSynapseIds;
+    } else {
+      // Find out which synapses were deleted and which were added
+      ({ onlyA: deletedSynapseIds, onlyB: addedSynapseIds } = diffArrays(
+        prevFilteredSynapseIds,
+        filteredSynapseIds,
+      ));
+    }
+
+    const layerName = segmentationLayer.name;
+    const skeleton = Store.getState().localSegmentationData[layerName].connectomeData.skeleton;
+    if (skeleton == null) return;
+
+    const { trees } = skeleton;
+
+    if (deletedSynapseIds.length > 0) {
+      const treeIdsToDelete = [];
+      deletedSynapseIds.forEach(synapseId =>
+        findTreeByName(trees, getTreeNameForSynapse(synapseId)).map(tree =>
+          treeIdsToDelete.push(tree.treeId),
+        ),
```

Comment:
Since is executed in bulk, one could theoretically compute a look up data structure from tree name to tree first. Right now, this performs a linear search for each synapse ID. I don't know how many synapse ids here are typically, so maybe it's completely unnecessary. Just crossed my mind when reading this..

Reply:
I had this in the back of my mind as a possible performance optimization, but didn't expect gains this large!

I loaded a moderately large segment (1421) with ~600 synapses and reset/filtered randomly. I measured the time to identify which trees to delete, given the IDs of the deleted synapses.

Before:

51.032958984375 ms
11.222900390625 ms
30.21484375 ms
36.15087890625 ms
0.9521484375 ms
38.079833984375 ms

After:

1.378662109375 ms
0.9638671875 ms
0.55908203125 ms
0.52685546875 ms
0.863037109375 ms
0.447998046875 ms

Thank you for bringing this up :partying_face:

Reply:
Wow, that's great news :)